### PR TITLE
kubernetes-cli: build for Apple Silicon

### PR DESCRIPTION
**Pull request by fxcoudert**
One of the most-downloaded formula with no Apple Silicon bottle… let's try to fix it

### CODE REVIEW

**Comment by fxcoudert** [comment]
```
==> make WHAT=cmd/kubectl
137
+++ [0107 07:34:15] Building go targets for darwin/arm64:
138
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
139
+++ [0107 07:34:15] Building go targets for darwin/arm64:
140
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
141
+++ [0107 07:34:15] Building go targets for darwin/arm64:
142
    ./vendor/github.com/go-bindata/go-bindata/go-bindata
143
+++ [0107 07:34:15] Building go targets for darwin/arm64:
144
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
145
+++ [0107 07:34:15] Building go targets for darwin/arm64:
146
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
147
+++ [0107 07:34:15] Building go targets for darwin/arm64:
148
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
149
touch: _output/bin/go-bindata: No such file or directory
150
```

**Comment by shastrula** [comment]
/opt brew install --build-from-source kubernetes-cli
==> Cloning https://github.com/kubernetes/kubernetes.git
Updating /Users/developer/Library/Caches/Homebrew/kubernetes-cli--git
==> Checking out tag v1.20.1
HEAD is now at c4d75276 Release commit for Kubernetes v1.20.1
HEAD is now at c4d75276 Release commit for Kubernetes v1.20.1
==> make WHAT=cmd/kubectl
Last 15 lines from /Users/developer/Library/Logs/Homebrew/kubernetes-cli/01.make:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
touch: _output/bin/go-bindata: No such file or directory
make[1]: *** [_output/bin/go-bindata] Error 1
make[1]: *** Waiting for unfinished jobs....
touch: _output/bin/deepcopy-gen: No such file or directory
make[1]: *** [_output/bin/deepcopy-gen] Error 1
touch: _output/bin/prerelease-lifecycle-gen: No such file or directory
make[1]: *** [_output/bin/prerelease-lifecycle-gen] Error 1
touch: _output/bin/defaulter-gen: No such file or directory
make[1]: *** [_output/bin/defaulter-gen] Error 1
touch: _output/bin/conversion-gen: No such file or directory
make[1]: *** [_output/bin/conversion-gen] Error 1
touch: _output/bin/openapi-gen: No such file or directory
make[1]: *** [_output/bin/openapi-gen] Error 1
make: *** [generated_files] Error 2

**Comment by thomasdemoor** [comment]
+1

**Comment by DrPsychick** [comment]
PR in progress - watch https://github.com/kubernetes/kubernetes/pull/97743

I was able to build it manually with the patch mentioned here: https://github.com/kubernetes/kubernetes/issues/97550#issue-775380450
...and copy it to `$HOMEBREW_PREFIX/bin/kubectl` - for the time being.

**Comment by chpshinmarq** [comment]
How is the progress for this issue?

**Comment by SMillerDev** [comment]
Support isn't merged upstream yet. So homebrew can't really support it either.

**Comment by kyranb** [comment]
Homebrew has the new version of Go now :)

**Comment by maxekman** [comment]
Can a rebuild be triggered? I'm sure this will build now when Go 1.16 is available.

**Comment by SMillerDev** [comment]
It needs to be rebased to retrigger

**Comment by Bo98** [comment]
> I'm sure this will build now

It won't without that aforementioned PR upstream.

**Comment by maxekman** [comment]
Upstream is now merged, not sure when it’ll be in a release though.

**Comment by Bo98** [comment]
1.21 will be released at the beginning of April I think. Given the patch is now approved and isn't too large, we could apply it as a patch here if it applies cleanly to 1.20.x.

**Comment by PPACI** [comment]
For people willing to have a little workaround to have It working right now.

1. Install a recent version of bash
2. Use it to compile kubectl locally

```shell
brew install bash
brew install --build-from-source --HEAD --env std kubernetes-cli
```


```
➜  ~ brew install --build-from-source --HEAD --env std kubernetes-cli
Warning: Calling brew install --env is deprecated! Use `env :std` in specific formula files instead.
==> Cloning https://github.com/kubernetes/kubernetes.git
Updating /Users/pierrepaci/Library/Caches/Homebrew/kubernetes-cli--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at c2ff2910fd7 Merge pull request #98797 from sxllwx/fix/issue98606
==> make WHAT=cmd/kubectl
==> hack/generate-docs.sh
==> Caveats
Bash completion has been installed to:
  /opt/homebrew/etc/bash_completion.d
==> Summary
🍺  /opt/homebrew/Cellar/kubernetes-cli/HEAD-c2ff291: 229 files, 52.7MB, built in 1 minute 58 seconds
```

On my MacBook Air M1

**Comment by SMillerDev** [comment]
> Warning: Calling brew install --env is deprecated! Use `env :std` in specific formula files instead.

This warning is here for a reason, if it needs a new bash version you're better off adding `depends_on "bash"` to your local copy before you run `brew install --build-from-source --HEAD kubernetes-cli`

**Comment by PPACI** [comment]
@SMillerDev You're absolutely right. 
But my snippet is more to be seen as an hacky temporary way to install it. At the second an official formula will be available, i'll drop my hack and switch to the real one.

**Comment by SMillerDev** [comment]
> But my snippet is more to be seen as an hacky temporary way to install it. At the second an official formula will be available, i'll drop my hack and switch to the real one.

It's not only that, it's also not likely to work unless people have an otherwise clean environment. If it needs bash as a dependency, add bash as a dependency.

**Comment by unitof** [comment]
_Does_ it depend on `bash` to build?

**Comment by PPACI** [comment]
Yes, it require a minimum of bash 4.1 to build. MacOS (at least mine…) contains bash 3.x by default.

**Comment by unitof** [comment]
I added the equivalent lines

```
  depends_on "bash" => :build
  env :std
```

to the `kubernetes-cli.rb` formula and indeed it builds! (with `--HEAD` flag only)

```
~ ▶ brew install --build-from-source kubernetes-cli --HEAD
==> Cloning https://github.com/kubernetes/kubernetes.git
Updating /Users/jacob/Library/Caches/Homebrew/kubernetes-cli--git
From https://github.com/kubernetes/kubernetes
   23d4b3b4f0d..1db614ec8fd  master     -> origin/master
==> Checking out branch master
Previous HEAD position was e87da0bd6e0 Release commit for Kubernetes v1.20.4
Switched to branch 'master'
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
HEAD is now at 1db614ec8fd Merge pull request #99472 from Huang-Wei/sched-bypass-init-pods-bench
==> make WHAT=cmd/kubectl
==> hack/generate-docs.sh
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions
==> Summary
🍺  /opt/homebrew/Cellar/kubernetes-cli/HEAD-1db614e: 229 files, 52.7MB, built in 2 minutes 13 seconds
Removing: /Users/jacob/Library/Caches/Homebrew/kubernetes-cli--1.20.2.big_sur.bottle.tar.gz... (12.2MB)
```

Genius, @PPACI!

It's unclear to me though if we should PR that change, or if it will no longer be necessary after next upstream release?

**Comment by PPACI** [comment]
For further reference for people coming here from google, you can edit the formula with `brew edit kubernetes-cli` and modify it like

```ruby
class KubernetesCli < Formula
  desc "Kubernetes command-line interface"
  homepage "https://kubernetes.io/"
  url "https://github.com/kubernetes/kubernetes.git",
      tag:      "v1.20.4",
      revision: "e87da0bd6e03ec3fea7933c4b5263d151aafd07c"
  license "Apache-2.0"
  head "https://github.com/kubernetes/kubernetes.git"

  livecheck do
    url :stable
    regex(/^v?(\d+(?:\.\d+)+)$/i)
  end

  bottle do
    sha256 cellar: :any_skip_relocation, big_sur:  "afb2863bbea76a02a679f613a8dff63014c7abee3ca10db76911a40d5512e2c4"
    sha256 cellar: :any_skip_relocation, catalina: "26cd1aaba450b81ae1f0536433f06ecd001be015b1d4cd2ee84ac0745176a505"
    sha256 cellar: :any_skip_relocation, mojave:   "20e9bc152b7e468fb2cc298934eac84fa606cf222d629deed4b6ad105eea3e90"
  end

  depends_on "go" => :build
  depends_on "bash" => :build
  env :std

...
```
(I've cropped the file for clarity, check the 2 last last lines of my snippet)
Then save and install it **the right way** as @unitof and @SMillerDev advised with `brew install --build-from-source kubernetes-cli --HEAD`

**Comment by unitof** [comment]
Turns out `env :std` is deprecated _altogether_, not just as a CLI command flag. So no to the PR.

```
~ ▶ brew audit --strict kubernetes-cli
kubernetes-cli:
  * 23: col 3: `env :std` in homebrew/core formulae is deprecated
Error: 1 problem in 1 formula detected
```

**Comment by carlocab** [comment]
Do you even need `env :std`? Does `brew install --HEAD` not work after adding `depends_on "bash" => :build` to the formula?

**Comment by unitof** [comment]
It didn’t for me. `env :std` was necessary.

**Comment by carlocab** [comment]
Probably means it's missing another build dependency that's present on your system but not in the build environment, because the formula is still missing another `depends_on`.

**Comment by ijoseph** [comment]
Anyone else getting `ln: /private/tmp/kubernetes-cli-20210314-82509-18zwz8/_output/bin/arm64: File exists`? 

```bash
 ❯ brew install --build-from-source --HEAD kubernetes-cli
==> Cloning https://github.com/kubernetes/kubernetes.git
Updating /Users/isaacjoseph/Library/Caches/Homebrew/kubernetes-cli--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at 854c2cc79f1 Merge pull request #100201 from aojea/flakekubeadm
==> make WHAT=cmd/kubectl
Last 15 lines from /Users/isaacjoseph/Library/Logs/Homebrew/kubernetes-cli/01.make:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
+++ [0314 20:38:24] Building go targets for darwin/arm64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
+++ [0314 20:38:24] Building go targets for darwin/arm64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
+++ [0314 20:38:24] Building go targets for darwin/arm64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
+++ [0314 20:38:24] Building go targets for darwin/arm64:
    ./vendor/github.com/go-bindata/go-bindata/go-bindata
ln: /private/tmp/kubernetes-cli-20210314-82509-18zwz8/_output/bin/arm64: File exists
!!! [0314 20:38:27] Call tree:
!!! [0314 20:38:27]  1: hack/make-rules/build.sh:28 kube::golang::place_bins(...)
make[1]: *** [_output/bin/prerelease-lifecycle-gen] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [generated_files] Error 2
```

?

**Comment by ijoseph** [comment]
Edit: worked again when I tried (1) in another `tmux` pane and (2) with `--env std` in the command line. Not sure which change fixed it. Looks like perhaps some sort of bizarre temp file collision, almost.

File: Formula/kubernetes-cli.rb [hunk 1]
```diff
@@ -48,8 +55,10 @@ def install
     run_output = shell_output("#{bin}/kubectl 2>&1")
     assert_match "kubectl controls the Kubernetes cluster manager.", run_output
 
-    version_output = shell_output("#{bin}/kubectl version --client 2>&1")
-    assert_match "GitTreeState:\"clean\"", version_output
+    # Due to ARM patch reinstate for next release
+    #version_output = shell_output("#{bin}/kubectl version --client 2>&1")
+    #assert_match "GitTreeState:\"clean\"", version_output
```
Comment (carlocab) [inline Formula/kubernetes-cli.rb:60]: Can we reference the test in the comment above the patch, so that this isn't forgotten when the patch is removed?

**Review by carlocab — APPROVED** [review]

**Pull request closed without merging**